### PR TITLE
Add corner_radius parameter for window and input box

### LINKED ISSUE #6: Layout configuration
For now layout is hardcoded and can only be change by changing code. These need to be configurable to fulfill the issue:

- [x] window position 5387c5fedae78451cd4598260684428355139c8a
- [x] window sizes b51997146bd4f30ab50461850312e03a6587260e
- [ ] arc radius of input block
- [x] margins of input block
- [x] margins of list as a whole
- [x] margins of list items

### LINKED ISSUE #41: Rounding of corners
image
You can see the edges of the popup of `yofi` are sharp. Can I get the option to make them rounded?? I don't mean the rounding of the input block, that's fine with me. I'm talking about the whole window/popup.

### PR DESCRIPTION
**Pull request by julianschuler**
This pull request adds the configuration parameter `corner_radius` for both the main application window as well as the input text field.

Minimal `yofi.config`:

```
corner_radius = "20"

[input_text]
corner_radius = "15 5 0 0"
```

The corner radius can be specified using one, two or four values with the same behavior as the CSS `border-radius` property (one value applies to all corners, two to opposite ones, and four to each corner separately in the order top-left, top-right, bottom-right, bottom-left).

The default behavior (no rounded corners for the application window, fully rounded ones for the input box) remains unchanged.

Resolves #6 and resolves #41.

### CODE REVIEW

File: src/draw/background.rs [hunk 1]
```diff
@@ -1,25 +1,45 @@
...
+        }
     }
 }
 
-impl Drawable for Background<'_> {
-    fn draw(self, dt: &mut DrawTarget<'_>, _: u16, _: Space, _: Point) -> Space {
-        dt.clear(self.params.color.as_source());
+impl Drawable for Background {
+    fn draw(self, dt: &mut DrawTarget<'_>, scale: u16, space: Space, start_point: Point) -> Space {
+        // dbg!(scale);
```
Comment (l4l) [inline src/draw/background.rs:30]: I guess it could be removed now

File: src/draw/background.rs [hunk 2]
```diff
@@ -1,25 +1,45 @@
...
+        // Clear the draw target to avoid artefacts for scales > 1 in the corners
+        dt.clear(SolidSource {
+            r: 0,
+            g: 0,
+            b: 0,
+            a: 0,
+        });
+
+        self.rect
+            .draw(dt, scale, Space { width, height }, start_point);
```
Comment (l4l) [inline src/draw/background.rs:42]: What's the point of the destructuring assignment earlier for immediate constructing the exact same type here?

File: src/style.rs [hunk 3]
```diff
@@ -176,3 +235,28 @@ impl FromStr for Margin {
...
+
+impl FromStr for Radius {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let values = s
+            .split(' ')
+            .map(|s| s.trim())
+            .filter(|s| !s.is_empty())
+            .map(|s| s.parse::<f32>().map_err(|_| "invalid float value"))
```
Comment (l4l) [inline src/style.rs:247]:
```suggestion
            .map(|s| s.parse::<f32>().map_err(|_| format!("invalid float value: {:?}", s)))
```

so no need to guess which one is incorrect

Comment (l4l) [inline src/style.rs:247]: It seems it is also possible to parse `f32::{INF,NEG_INF,NAN}` which barely make sense there. So does negative values

File: src/config.rs [hunk 4]
```diff
@@ -70,6 +72,8 @@ struct InputText {
     margin: Margin,
     #[def = "Padding::from_pair(1.7, -4.0)"]
     padding: Padding,
+    #[def = "Radius::all(f32::INFINITY)"]
```
Comment (l4l) [inline src/config.rs:75]: It's not really an infinity, isn't it?

File: src/draw.rs [hunk 5]
```diff
@@ -67,3 +72,70 @@ where
...
+        Self { radius, color }
+    }
+}
+
+impl Drawable for RoundedRect {
+    fn draw(self, dt: &mut DrawTarget<'_>, scale: u16, space: Space, start_point: Point) -> Space {
+        let Point { x, y, _unit } = start_point;
+        let Space { width, height } = space;
+
+        let max_radius = width.min(height) / 2.0;
```
Comment (l4l) [inline src/draw.rs:92]: This one looks like an artificial restriction. Why is that needed?

**Review by l4l — COMMENTED** [review]
This looks really great so far!

File: src/draw/background.rs (hunk 1 above)
Comment (julianschuler) [reply src/draw/background.rs]: Of course, fixed.

File: src/config.rs (hunk 4 above)
Comment (julianschuler) [reply src/config.rs]: You're right, `f32::MAX` is probably more appropriate.

File: src/draw/background.rs (hunk 2 above)
Comment (julianschuler) [reply src/draw/background.rs]: There is none (anymore), fixed.

File: src/draw.rs (hunk 5 above)
Comment (julianschuler) [reply src/draw.rs:92]: The idea here is to employ a similar principle as for the CSS border-radius property: We don't want any corner curves to overlap and thus cap the radius to the above described one. This results in semi-circles at the smaller side for too large radii (like for the default input text field).

While this is not the exact same behavior as for CSS ([it specifies to resizes all radii proportionally](https://www.w3.org/TR/css-backgrounds-3/#corner-overlap)), it should be close enough and allows specifying large values to force maximum rounding (to semi-circles).

Please let me know if I should change the implementation to the one specified by CSS.

File: src/style.rs (hunk 3 above)
Comment (julianschuler) [reply src/style.rs:247]: 👍🏾

File: src/draw.rs (hunk 5 above)
Comment (l4l) [reply src/draw.rs:92]: Oh, that's interesting. I don't think the complication of proportional resizing really worth it so let it be like this. But a small comment here might be helpful there

Comment (julianschuler) [reply src/draw.rs:92]: Alright, thank you!

File: src/style.rs (hunk 3 above)
Comment (julianschuler) [reply src/style.rs]: Good catch. I have now added a check prohibiting parsing non-finite float values for `Radius`, `Margin` and `Padding` as well as one allowing only positive values for `Radius`.

**Comment by l4l** [comment]
Thanks for your contribution! I was really lazy to handle that one. Hopefully would release this within this/next week

**Comment by julianschuler** [comment]
Awesome! Also, thanks for your review!
